core/deployment/src/main/java/io/quarkus/deployment/dev/DevModeContext.java

```diff
@@ -422,10 +430,11 @@ public CompilationUnit(PathCollection sourcePaths, String classesPath, PathColle
             this.classesPath = classesPath;
             this.resourcePaths = resourcePaths;
             this.resourcesOutputPath = resourcesOutputPath;
+            this.precedingSourcePaths = PathList.empty();
         }
 
         public PathCollection getSourcePaths() {
-            return sourcePaths;
+            return sourcePaths.addAllFirst(precedingSourcePaths);
```

Comment:
It'd be better to add them in the constructor to avoid re-adding them on every call. There is code that's doing `getSourcePaths().isEmpty()` and then calls it again to iterate.